File: node-server/src/services/file-service.ts
import { readFile, readdir, realpath, stat } from "node:fs/promises";
import { extname, relative, resolve } from "node:path";

import { ApiError } from "../errors.js";
import { WorkspaceService } from "./workspace-service.js";

const IGNORED_NAMES = new Set(["node_modules", ".git", ".next", "dist", "build", "__pycache__", ".turbo", ".cache", "coverage", ".pytest_cache", ".mypy_cache", "target", "vendor", ".ds_store"]);
const SENSITIVE_DIRECTORIES = new Set([".ssh", ".aws", ".azure", ".gnupg"]);
const SENSITIVE_NAMES = new Set([".env", "secrets.env", "credentials", "credentials.json"]);
const SENSITIVE_SUFFIXES = new Set([".pem", ".key", ".p12", ".pfx"]);
const TEXT_LIMIT = 256 * 1024;
const MEDIA_LIMIT = 10 * 1024 * 1024;

const LANGUAGES: Record<string, string> = {
  ".ts": "typescript", ".tsx": "typescript", ".js": "javascript", ".jsx": "javascript", ".mjs": "javascript", ".cjs": "javascript",
  ".py": "python", ".go": "go", ".rs": "rust", ".java": "java", ".html": "html", ".css": "css", ".json": "json",
  ".yaml": "yaml", ".yml": "yaml", ".toml": "toml", ".md": "markdown", ".sh": "bash", ".sql": "sql",
};

export interface FileEntry {
  name: string;
  isDir: boolean;
  size: number;
  modified: string;
}

export class FileService {
  constructor(private readonly workspaces: WorkspaceService) {}

  async list(path: string, root: string): Promise<{ entries: FileEntry[]; path: string }> {
    const target = await this.resolve(path, root);
    const targetStat = await this.fileStat(target, "Directory");
    if (!targetStat.isDirectory()) throw new ApiError(400, "not_a_directory", "Path is not a directory");
    const entries = await readdir(target, { withFileTypes: true });
    const result = await Promise.all(entries
      .filter((entry) => !this.ignored(entry.name))
      .map(async (entry) => {
        try {
          const item = await stat(resolve(target, entry.name));
          return {
            name: entry.name,
            isDir: entry.isDirectory(),
            size: entry.isFile() ? item.size : 0,
            modified: item.mtime.toISOString(),
          };
        } catch {
          return undefined;
        }
      }));
    return {
      entries: result.filter((entry): entry is FileEntry => entry !== undefined)
        .sort((left, right) => Number(right.isDir) - Number(left.isDir) || left.name.localeCompare(right.name)),
      path: target,
    };
  }

  async readText(path: string, root: string): Promise<{ content: string; language: string; size: number }> {
    const target = await this.resolve(path, root);
    const info = await this.fileStat(target, "File");
    if (!info.isFile()) throw new ApiError(400, "not_a_file", "Path is not a file");
    if (info.size > TEXT_LIMIT) throw new ApiError(413, "file_too_large", `Text preview exceeds ${TEXT_LIMIT} bytes`);
    const content = await readFile(target, "utf8").catch((error: NodeJS.ErrnoException) => {
      if (error.code === "ERR_INVALID_ARG_VALUE") throw error;
      throw new ApiError(415, "binary_file", "File is not valid UTF-8 text");
    });
    return { content, language: this.language(target), size: info.size };
  }

  async media(path: string, root: string): Promise<{ content: Buffer; mimeType: string }> {
    const target = await this.resolve(path, root);
    const info = await this.fileStat(target, "File");
    if (!info.isFile()) throw new ApiError(400, "not_a_file", "Path is not a file");
    const mimeType = this.mimeType(target);
    if (mimeType === undefined) throw new ApiError(415, "unsupported_media", "File is not previewable media");
    if (info.size > MEDIA_LIMIT) throw new ApiError(413, "file_too_large", `Media preview exceeds ${MEDIA_LIMIT} bytes`);
    return { content: await readFile(target), mimeType };
  }

  private async resolve(path: string, rootInput: string): Promise<string> {
    const roots = await this.workspaces.roots();
    let root: string;
    try { root = await realpath(rootInput); } catch { throw new ApiError(403, "root_not_allowed", "Workspace root is not allowed"); }
    if (!roots.some((item) => this.samePath(item, root))) {
      throw new ApiError(403, "root_not_allowed", "Workspace root is not allowed");
    }
    const candidate = resolve(root, path || ".");
    const candidateRemainder = relative(root, candidate);
    if (candidateRemainder.startsWith("..") || candidateRemainder.includes(":")) {
      throw new ApiError(403, "path_outside_workspace", "Path escapes the selected workspace");
    }
    let target: string;
    try { target = await realpath(candidate); } catch { throw new ApiError(404, "file_not_found", "File was not found"); }
    const remainder = relative(root, target);
    if (remainder.startsWith("..") || remainder.includes(":")) {
      throw new ApiError(403, "path_outside_workspace", "Path escapes the selected workspace");
    }
    if (this.sensitive(remainder)) throw new ApiError(403, "sensitive_file", "Sensitive files cannot be previewed");
    return target;
  }

  private async fileStat(path: string, label: string) {
    try { return await stat(path); } catch { throw new ApiError(404, "file_not_found", `${label} was not found`); }
  }

  private samePath(left: string, right: string): boolean { return left.toLocaleLowerCase() === right.toLocaleLowerCase(); }
  private ignored(name: string): boolean {
    const lower = name.toLocaleLowerCase();
    return IGNORED_NAMES.has(lower) || SENSITIVE_NAMES.has(lower) || lower.startsWith(".env.") || lower.endsWith(".pyc") || SENSITIVE_SUFFIXES.has(extname(lower));
  }
  private sensitive(relativePath: string): boolean {
    const parts = relativePath.split(/[\\/]/).map((part) => part.toLocaleLowerCase());
    const name = parts.at(-1) ?? "";
    return parts.some((part) => SENSITIVE_DIRECTORIES.has(part)) || SENSITIVE_NAMES.has(name) || name.startsWith(".env.") || SENSITIVE_SUFFIXES.has(extname(name)) || (name === "config" && parts.includes(".git"));
  }
  private language(path: string): string { return LANGUAGES[extname(path).toLocaleLowerCase()] ?? "text"; }
  private mimeType(path: string): string | undefined {
    const extension = extname(path).toLocaleLowerCase();
    const image: Record<string, string> = { ".png": "image/png", ".jpg": "image/jpeg", ".jpeg": "image/jpeg", ".gif": "image/gif", ".webp": "image/webp", ".svg": "image/svg+xml", ".bmp": "image/bmp", ".ico": "image/x-icon" };
    const audio: Record<string, string> = { ".mp3": "audio/mpeg", ".wav": "audio/wav", ".ogg": "audio/ogg", ".m4a": "audio/mp4", ".aac": "audio/aac", ".flac": "audio/flac" };
    return image[extension] ?? audio[extension];
  }
}

File: pi-python/server/services/file_service.py
"""Workspace-scoped file browsing and preview operations.

中文说明：文件服务：目录浏览、文本/媒体预览与变化监听。
所有访问先校验根目录白名单，再做真实路径边界与敏感文件检查。
"""

from __future__ import annotations

import asyncio
from collections.abc import AsyncIterator, Callable, Iterable
from datetime import datetime, timezone
import json
import mimetypes
from pathlib import Path
from typing import Any


IGNORED_NAMES = {
    # 目录浏览时忽略的重型/生成目录
    "node_modules",
    ".git",
    ".next",
    "dist",
    "build",
    "__pycache__",
    ".turbo",
    ".cache",
    "coverage",
    ".pytest_cache",
    ".mypy_cache",
    "target",
    "vendor",
    ".DS_Store",
}
IGNORED_SUFFIXES = {".pyc"}
SENSITIVE_DIRECTORY_NAMES = {".ssh", ".aws", ".azure", ".gnupg"}
# 禁止预览的敏感文件/后缀（.env、密钥、凭据等）
SENSITIVE_FILE_NAMES = {".env", "secrets.env", "credentials", "credentials.json"}
SENSITIVE_SUFFIXES = {".pem", ".key", ".p12", ".pfx"}
TEXT_PREVIEW_MAX_BYTES = 256 * 1024
MEDIA_PREVIEW_MAX_BYTES = 10 * 1024 * 1024

EXT_TO_LANGUAGE = {
    "ts": "typescript",
    "tsx": "typescript",
    "js": "javascript",
    "jsx": "javascript",
    "mjs": "javascript",
    "cjs": "javascript",
    "py": "python",
    "rb": "ruby",
    "go": "go",
    "rs": "rust",
    "java": "java",
    "kt": "kotlin",
    "swift": "swift",
    "c": "c",
    "cpp": "cpp",
    "h": "c",
    "hpp": "cpp",
    "cs": "csharp",
    "html": "html",
    "htm": "html",
    "css": "css",
    "scss": "css",
    "less": "css",
    "json": "json",
    "jsonl": "json",
    "yaml": "yaml",
    "yml": "yaml",
    "toml": "toml",
    "xml": "xml",
    "md": "markdown",
    "mdx": "markdown",
    "sh": "bash",
    "bash": "bash",
    "zsh": "bash",
    "sql": "sql",
    "graphql": "graphql",
    "gql": "graphql",
    "tf": "hcl",
    "hcl": "hcl",
    "gitignore": "bash",
    "txt": "text",
}


class FileAccessError(Exception):
    """文件访问错误：状态码 + 机器码 + 消息，路由转成 APIError。"""
    def __init__(self, status_code: int, code: str, message: str) -> None:
        super().__init__(message)
        self.status_code = status_code
        self.code = code
        self.message = message


class FileService:
    """工作区文件操作：roots_provider 动态提供允许的根目录集合。"""
    def __init__(self, roots_provider: Callable[[], Iterable[str | Path]]) -> None:
        self._roots_provider = roots_provider

    def allowed_roots(self) -> tuple[Path, ...]:
        """解析并去重当前允许的根目录。"""
        unique: dict[str, Path] = {}
        for root in self._roots_provider():
            resolved = Path(root).expanduser().resolve()
            unique[_path_key(resolved)] = resolved
        return tuple(unique.values())

    def resolve(self, file_path: str, root_path: str) -> tuple[Path, Path]:
        """解析目标路径：根目录必须白名单内，解析后必须仍位于根内，且不是敏感文件。"""
        root = Path(root_path).expanduser().resolve()
        allowed = {_path_key(item) for item in self.allowed_roots()}
        if _path_key(root) not in allowed:
            raise FileAccessError(403, "root_not_allowed", "Workspace root is not allowed")

        candidate = Path(file_path).expanduser()
        if not candidate.is_absolute():
            candidate = root / candidate
        resolved = candidate.resolve()
        try:
            resolved.relative_to(root)
        except ValueError as exception:
            raise FileAccessError(
                403,
                "path_outside_workspace",
                "Path escapes the selected workspace",
            ) from exception
        if _is_sensitive(resolved, root):
            raise FileAccessError(403, "sensitive_file", "Sensitive files cannot be previewed")
        return resolved, root

    def list_directory(self, file_path: str, root_path: str) -> dict[str, Any]:
        """列目录：过滤忽略项，返回名称/类型/大小/修改时间，目录优先排序。"""
        target, _ = self.resolve(file_path, root_path)
        if not target.exists():
            raise FileAccessError(404, "file_not_found", "Directory was not found")
        if not target.is_dir():
            raise FileAccessError(400, "not_a_directory", "Path is not a directory")
        entries: list[dict[str, Any]] = []
        for child in target.iterdir():
            if _should_ignore(child):
                continue
            try:
                stat = child.stat()
            except OSError:
                continue
            entries.append(
                {
                    "name": child.name,
                    "isDir": child.is_dir(),
                    "size": stat.st_size if child.is_file() else 0,
                    "modified": datetime.fromtimestamp(
                        stat.st_mtime,
                        timezone.utc,
                    ).isoformat(),
                }
            )
        entries.sort(key=lambda item: (not item["isDir"], item["name"].casefold()))
        return {"entries": entries, "path": str(target)}

    def read_text(self, file_path: str, root_path: str) -> dict[str, Any]:
        """读文本预览：限制大小与 UTF-8 编码，附带语法高亮语言。"""
        target, _ = self.resolve(file_path, root_path)
        stat = self._file_stat(target)
        if stat.st_size > TEXT_PREVIEW_MAX_BYTES:
            raise FileAccessError(
                413,
                "file_too_large",
                f"Text preview exceeds {TEXT_PREVIEW_MAX_BYTES} bytes",
            )
        try:
            content = target.read_text(encoding="utf-8")
        except UnicodeDecodeError as exception:
            raise FileAccessError(
                415,
                "binary_file",
                "File is not valid UTF-8 text",
            ) from exception
        return {
            "content": content,
            "language": guess_language(target),
            "size": stat.st_size,
        }

    def media_file(self, file_path: str, root_path: str) -> tuple[Path, str]:
        """返回可预览的媒体文件（图片/音频）及其 MIME。"""
        target, _ = self.resolve(file_path, root_path)
        stat = self._file_stat(target)
        mime = mimetypes.guess_type(target.name)[0] or "application/octet-stream"
        if not (mime.startswith("image/") or mime.startswith("audio/")):
            raise FileAccessError(415, "unsupported_media", "File is not previewable media")
        if stat.st_size > MEDIA_PREVIEW_MAX_BYTES:
            raise FileAccessError(
                413,
                "file_too_large",
                f"Media preview exceeds {MEDIA_PREVIEW_MAX_BYTES} bytes",
            )
        return target, mime

    async def watch(
        self,
        file_path: str,
        root_path: str,
        *,
        poll_seconds: float = 1,
    ) -> AsyncIterator[str]:
        """轮询监听文件变化，通过 SSE 推送 change 事件（附带心跳）。"""
        target, _ = self.resolve(file_path, root_path)
        stat = self._file_stat(target)
        previous = (stat.st_mtime_ns, stat.st_size)
        yield _named_sse("connected", {"filePath": str(target)})
        while True:
            await asyncio.sleep(max(0.1, poll_seconds))
            try:
                stat = target.stat()
                current = (stat.st_mtime_ns, stat.st_size)
            except OSError:
                current = (0, 0)
            if current == previous:
                yield ": heartbeat\n\n"
                continue
            previous = current
            yield _named_sse(
                "change",
                {
                    "mtime": datetime.now(timezone.utc).isoformat(),
                    "size": current[1],
                },
            )

    @staticmethod
    def _file_stat(target: Path):
        if not target.exists():
            raise FileAccessError(404, "file_not_found", "File was not found")
        if not target.is_file():
            raise FileAccessError(400, "not_a_file", "Path is not a file")
        return target.stat()


def guess_language(path: Path) -> str:
    """按文件名/后缀猜测代码语言，用于前端高亮。"""
    base = path.name.casefold()
    if base == "dockerfile" or base.startswith("dockerfile."):
        return "dockerfile"
    if base in {"makefile", "gnumakefile"}:
        return "makefile"
    return EXT_TO_LANGUAGE.get(path.suffix.casefold().removeprefix("."), "text")


def _path_key(path: Path) -> str:
    return str(path).casefold()


def _is_sensitive(path: Path, root: Path) -> bool:
    """敏感判定：敏感目录、.env/密钥文件、凭据后缀、.git/config 等。"""
    relative_parts = [part.casefold() for part in path.relative_to(root).parts]
    if any(part in SENSITIVE_DIRECTORY_NAMES for part in relative_parts):
        return True
    name = path.name.casefold()
    return (
        name in SENSITIVE_FILE_NAMES
        or name.startswith(".env.")
        or path.suffix.casefold() in SENSITIVE_SUFFIXES
        or (name == "config" and ".git" in relative_parts)
    )


def _should_ignore(path: Path) -> bool:
    """目录浏览时是否忽略：重型目录、敏感文件、pyc/密钥后缀。"""
    name = path.name.casefold()
    return (
        name in {item.casefold() for item in IGNORED_NAMES}
        or name in SENSITIVE_FILE_NAMES
        or name.startswith(".env.")
        or path.suffix.casefold() in IGNORED_SUFFIXES | SENSITIVE_SUFFIXES
    )


def _named_sse(event: str, data: dict[str, Any]) -> str:
    payload = json.dumps(data, ensure_ascii=False, separators=(",", ":"))
    return f"event: {event}\ndata: {payload}\n\n"
